api/queries_pr.go

```diff
@@ -268,38 +287,133 @@ func (pr *PullRequest) ChecksStatus() PullRequestChecksStatus {
 		return summary
 	}
 
-	commit := pr.StatusCheckRollup.Nodes[0].Commit
-	for _, c := range commit.StatusCheckRollup.Contexts.Nodes {
-		// Nodes are a discriminated union of CheckRun or StatusContext,
-		// but we can use the State field to disambiguate, rather than using the TypeName.
-		// First we try to get the State, as if we have a StatusContext
-		state := c.State
-		// But if the State is empty, then we assume we actually have a CheckRun
-		if state == "" {
-			// If the Status of a CheckRun is COMPLETED, then we want to look more closely
-			// at the Conclusion, as that contains the relevant information.
+	contexts := pr.StatusCheckRollup.Nodes[0].Commit.StatusCheckRollup.Contexts
+
+	// If this commit has counts by state then we can summarise check status from those
+	if len(contexts.CheckRunCountsByState) != 0 && len(contexts.StatusContextCountsByState) != 0 {
+		summary.Total = contexts.CheckRunCount + contexts.StatusContextCount
+		for _, countByState := range contexts.CheckRunCountsByState {
+			switch parseCheckStatusFromCheckRunState(countByState.State) {
+			case passing:
+				summary.Passing += countByState.Count
+			case failing:
+				summary.Failing += countByState.Count
+			default:
+				summary.Pending += countByState.Count
+			}
+		}
+
+		for _, countByState := range contexts.StatusContextCountsByState {
+			switch parseCheckStatusFromStatusState(countByState.State) {
+			case passing:
+				summary.Passing += countByState.Count
+			case failing:
+				summary.Failing += countByState.Count
+			default:
+				summary.Pending += countByState.Count
+			}
+		}
+
+		return summary
+	}
+
+	// If we don't have the counts by state, then we'll need to summarise by looking at the more detailed contexts
+	for _, c := range contexts.Nodes {
+		// Nodes are a discriminated union of CheckRun or StatusContext and we can match on
+		// the TypeName to narrow the type.
+		if c.TypeName == "CheckRun" {
+			// https://docs.github.com/en/graphql/reference/enums#checkstatusstate
+			// If the status is completed then we can check the conclusion field
 			if c.Status == "COMPLETED" {
-				state = c.Conclusion
+				switch parseCheckStatusFromCheckConclusionState(c.Conclusion) {
```

Comment:
I was finding it not the easiest to keep the previous flow in my head, and found myself frequently referring back to the graphql docs to determine whether I was dealing with a `CheckStatusState`, a `StatusState`, a `CheckRunState`, a `CheckConclusionState` or something that might exist in more than one category.

Therefore, I decided to eschew any attempt to treat these things as the same and to parse them each separately. One nice result of this is that when we are able to remove the logic for `StatusContext` it'll be very easy to carve out the unneeded code.

Reply:
One last thing I'm interesting in exploring here is using the enum support in shurcool to make these `parse` functions a bit more type-friendly (as opposed to stringly typed)